### PR TITLE
feat: support accessing other fixtures in fixture function

### LINKED ISSUE #3626: Support accessing other fixtures in the fixture function when using `test.extend`
### Clear and concise description of the problem

This idea comes from the [question](https://github.com/vitest-dev/vitest/pull/3554#issuecomment-1597344051) by @outerlook. 

When setting up fixtures in `test.extend`,  we may dependent on other fixtures. Would be very useful to have this feature.

```ts
export const newTest = test.extend({
  server: async (use) => {
    const server = await startServer();
    await use(server);

    await server.destroy();
  },
  // client depends on server
  client: async (use, { server }) => {
    const client = await new Client({ apiUrl: server.apiUrl });
    await use(client);

    await client.destroy();
  }
});
```

### Suggested solution

Make other fixtures accessible as the second argument of each fixture function.

### Alternative

_No response_

### Additional context

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by fenghan34**
Closes #3626

### CODE REVIEW

File: packages/runner/src/fixture.ts [hunk 1]
```diff
@@ -1,66 +1,162 @@
...
 
-      const colonIndex = _prop.indexOf(':')
-      if (colonIndex > 0)
-        _prop = _prop.slice(0, colonIndex).trim()
+  return pendingFixtures
+}
+
+function getFixtureFnDepProps(fn: Function, allProps: string[]) {
+  if (fn.length !== 2)
+    throw new Error('fixture function should have two arguments, the fist one is the use function that should be called with fixture value, and the second is other fixtures that should be used with destructured expression. For example, `async ({ a, b }, use) => { await use(a + b) }`')
```
Comment (sheremet-va) [inline packages/runner/src/fixture.ts:107]: This looks misleading 😄

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: packages/runner/src/fixture.ts (hunk 1 above)
Comment (fenghan34) [reply packages/runner/src/fixture.ts:107]: Aha, I made a mistake. 😅

**Comment by sheremet-va** [comment]
I wonder if it would be a good idea to support the same arguments order, so we also support third-party playwright plugins?

**Comment by sheremet-va** [comment]
Also, would be nice to release this in the next version alongside the support for `test.extend`. Since we didn't release it yet, changing arguments order is not a breaking change.

**Comment by fenghan34** [comment]
> I wonder if it would be a good idea to support the same arguments order, so we also support third-party playwright plugins?

We can do that, but I think it's kind of strict that users must follow this pattern `async ({}, use) => {}` to declare a fixture function. Btw, what are third-party playwright plugins?

**Comment by fenghan34** [comment]
In Playwright, users must use the object destructuring pattern `{a, b, c}` to access fixtures through the first argument both in the test and fixture function, the default function parameters `({ a = 1 }) => {}` and rest parameters `({ a, b, ...rest }) => {}` are not allowed. 

I prefer to make it more flexible with mentioned features above, so the following test/fixture function declarations would work with the current implementation but not Playwright. 

```ts
// Be aware the second argument is the `use` function in Playwright 
const myTest = test.extend({
  a: async (use) => {
    await use()
  },
  // b depends on a and c
  b: async (use, context) => {
    await use()
  },
  // c depends on a and b
  c: async (...args) => {
  	await args[0]()
  },
})

const myTest2 = myTest.extend({
  // d depends on a
  d: async (use, { a = 1 }) => {
  	await use()
  },
  // e depends on a, b, c and d
  e: async (use, { a, ...rest }) => {
    await use()
  }
})

myTest2('', (context) => {}) // all fixtures will be initialized
myTest2('', (...args) => {}) // all fixtures will be initialized
myTest2('', ({ a, ...rest }) => {}) // all fixtures will be initialized
myTest2('', ({ d = 1 }) => {}) // only a and d will be initialized
```

This gives users more convenience but also risks of circular dependency when declaring fixture functions with `(...rest) => {}`, `(use, { a, ...rest}) => {}` or `(use, context) => {}`. Would like to hear your thoughts 😄.

**Comment by sheremet-va** [comment]
> Would like to hear your thoughts 😄.

I think it's better to be more strict. We don't know the order of fixtures that we need to call to support `context` without destructuring, and I don't see big benefits for this pattern - it's better to use only things you need without initializing _everything_ (even things you don't rely on).

> the default function parameters

It doesn't make sense to support a default value here.

> what are third-party playwright plugins?

Any package that uses `test.extend` under the hood.

**Comment by fenghan34** [comment]
> I think it's better to be more strict. We don't know the order of fixtures that we need to call to support `context` without destructuring, and I don't see big benefits for this pattern - it's better to use only things you need without initializing _everything_ (even things you don't rely on).

I see. Make sense to me. So do we need to restrict the test function(destructuring context) as well?

**Comment by sheremet-va** [comment]
> I see. Make sense to me. So do we need to restrict the test function(destructuring context) as well?

I would think so

**Comment by fenghan34** [comment]
> I would think so

Got it, thank you for replying, will push more commits.

**Comment by sheremet-va** [comment]
@fenghan34 any news? We are planning to make a release today 👀

**Comment by fenghan34** [comment]
> @fenghan34 any news? We are planning to make a release today 👀 

I'm writing tests，will update it today😄

File: docs/guide/test-context.md [hunk 2]
```diff
@@ -115,15 +115,17 @@ const myTest = test.extend({
...
 
 // todosFn will run
-myTest('', ({ todos }) => {}) // only todos is available
-myTest('', (context) => {}) // both are available
-myTest('', ({ archive, ...rest }) => {}) // both are available
+myTest('', ({ todos }) => {})
 ```
 
+::: warning
+When using `test.extend()`, you should always use the object destructuring pattern `{ todos }` to access context both in fixture function and test function.
```
Comment (sheremet-va) [inline docs/guide/test-context.md:126]: I think we should allow normal access (without destructuring) if there are no fixtures added with `test.extend` for backwards compatibility

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: docs/guide/test-context.md (hunk 2 above)
Comment (fenghan34) [reply docs/guide/test-context.md:126]: Make sense

**Review by sheremet-va — APPROVED** [review]

**Pull request merged**